### PR TITLE
Ascii art

### PR DESCRIPTION
**Pull request by hdhoang**

### CODE REVIEW

**Comment by hdhoang** [comment]
Tomorrow I can unpin nightly in this PR too, to avoid too much churn.

File: src/layout.rs [hunk 1]
```diff
@@ -2,20 +2,6 @@ use action::Action;
...
-  |-----------------------------------------------------------------------------|
-  |Tab    |  Q  |  W  |  E  |  R  |  T  |  Y  |  U  |  I|   O|  P|  [|  ]|  \ ] |
-  |-----------------------------------------------------------------------------|
-  |Caps         |    A|    S|    D|    F|   G|  H|  J|  K|  L|  ;|  '|   #|Enter|
-  |-----------------------------------------------------------------------------|
-  |Shift      |    Z|     X|    C|     V|  B|  N|  M|  ,|  .|  /|     Shift     |
-  |-----------------------------------------------------------------------------|
-  |Ctrl |Meta | Alt |               Space                |Alt | Fn  | Anne |Ctrl|
-  `-----------------------------------------------------------------------------'
-*/
```
Comment (ah-) [inline src/layout.rs]: I kinda liked that picure :(

File: src/layout.rs [hunk 2]
```diff
@@ -27,33 +13,53 @@ const BT_T: Action = LayerToggle(3);
...
+// Row 1:
+    Escape  ;  N1 ;  N2 ;  N3 ;  N4 ;  N5 ;  N6 ;  N7 ;  N8 ;  N9 ;  N0 ; Minus ; Equal ;        BSpace;
+// Row 2:
+    Tab      ;  Q  ;  W  ;  E  ;  R  ;  T  ;  Y  ;  U  ;  I  ;  O  ;  P  ; LBracket ; RBracket ; BSlash;
+// Row 3:
+    Capslock  ;  A  ;  S  ;  D  ;  F  ;  G  ;  H  ;  J  ;  K  ;  L  ; SColon ; Quote ;         No;Enter;
+// Row 4:
+    LShift     ;  Z  ;  X  ;  C  ;  V  ;  B  ;  N  ;  M  ; Comma ; Dot ; Slash ;           No;No;RShift;
+// Row 5:
+    LCtrl ; LMeta ; LAlt ;  No;No;      Space ; No;No;No;No;  RAlt  ; FN_M  ;  BT_T  ;  LED_M
```
Comment (ah-) [inline src/layout.rs]: Hmm, to be honest this is a bit harder to read than it was before?

**Comment by ah-** [comment]
+1 for the emacs gitignore, I like that.

Something like the below (with better key alignment, was lazy) would be fun macro abuse, but I'm not quite sure if it's worth spending the time on:

```
pub const BASE: Layout = layout![
  ,-----------------------------------------------------------------------------------------------------------.
  | Escape   | N1    | N2   | N3 | N4 | N5    | N6 | N7 | N8    | N9  | N0     | Minus    | Equal    | BSpace |
  |-----------------------------------------------------------------------------------------------------------|
  | Tab      | Q     | W    | E  | R  | T     | Y  | U  | I     | O   | P      | LBracket | RBracket | BSlash |
  |-----------------------------------------------------------------------------------------------------------|
  | Capslock      | A   | S   | D   | F   | G   | H   | J   | K   | L    | SColon | Quote   |          Enter  |
  |-----------------------------------------------------------------------------------------------------------|
  | LShift     | Z     | X    | C    | V    | B     | N    | M    | Comma | Dot | Slash  |             RShift |
  |-----------------------------------------------------------------------------------------------------------|
  | LCtrl    | LMeta | LAlt |                Space                    | RAlt   | FN_M     | BT_T     | LED_M  |
  `-----------------------------------------------------------------------------------------------------------'
];
```

In practice it's probably not worth it, with longer and more meaningful Action codes it'll all shift around anyways.

**Comment by hdhoang** [comment]
I will close this PR, and split out the gitignore bit. Do you think `book.*` should be skipped by  CI? That can go along.

The main hurdle I had with ASCII-art characters is that they are matched by `expr`. The name-only keys/actions can be matched with `$key: ident`, and I'll try adding an `$action:ident ( $value:tt )` pattern to cover for the struct actions.

**Comment by ah-** [comment]
I've got a whole list of To-dos at home, will post some tonight.

**Comment by ah-** [comment]
I've added a bunch of them to https://github.com/ah-/anne-key/issues, hope there's something interesting in there for you.

Some of them like https://github.com/ah-/anne-key/issues/27 are pretty doable (and testable) with just Rust.

**Pull request closed without merging**